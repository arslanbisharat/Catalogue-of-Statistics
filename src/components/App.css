.App {
  text-align: center;
}

.app-title {
  margin-top: 10vh;
  font-size: 4rem;
}

.app-button {
  background-color: black;
  color: white;
  width: 30%;
  height: 25%;
  font-size: 50px;
  font-family: 'batfont';
  margin-top: 10vh;
}