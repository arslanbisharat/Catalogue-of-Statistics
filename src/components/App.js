import React, { useState, useEffect } from 'react';
import {
  BrowserRouter as Router,
  Switch,
  Route,
} from 'react-router-dom';
import axios from 'axios';
import PropTypes from 'prop-types';
import HeroesList from '../containers/HeroesList';
import HeroFile from './HeroFile';
import TeamsList from '../containers/TeamsList';
import Home from './Home';
import About from './About';
import './App.css';

const App = props => {
  const [heros, setHeros] = useState();
  const {
    chngFilter, heroes, filter, addNewHero,
  } = props;

  const handleAddHero = () => {
    if (heros) {
      heros.heroes.forEach(value => {
        addNewHero(value);
      });
    }
  };

  useEffect(() => {
    let mounted = true;

    const loadHeroes = async () => {
      const response = await axios.get('https://superheroapi.com/api.php/2602016920125978/search/_');

      if (mounted) {
        setHeros({
          heroes: response.data.results.filter(heroes => heroes.biography.publisher === 'DC Comics'
            || heroes.biography.publisher === 'Superman Prime One-Million'
            || heroes.biography.publisher === 'Black Racer'),
        });

        handleAddHero();
      }
    };

    loadHeroes();

    return () => {
      mounted = false;
    };
  }, heros);

  const clickHandler = value => {
    chngFilter(value);
  };

  return (
    <Router>
      <div className="App">
        <Switch>
          <Route exact path="/">
            <Home />
          </Route>
          <Route path="/heroesList">
            <HeroesList />
          </Route>
          <Route path="/teamsList">
            <TeamsList />
          </Route>
          <Route path="/heroFile/:id">
            <HeroFile
              hero={heroes.filter(hero => hero.id === filter.value[1])[0]}
              filter={filter}
              handleClick={clickHandler}
            />
          </Route>
          <Route path="/about">
            <About />
          </Route>
        </Switch>
      </div>
    </Router>
  );
};

App.propTypes = {
  addNewHero: PropTypes.func.isRequired,
  chngFilter: PropTypes.func.isRequired,
  heroes: PropTypes.arrayOf(PropTypes.object).isRequired,
  filter: PropTypes.shape({
    value: PropTypes.arrayOf(PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
    ])),
  }).isRequired,
};

export default App;
