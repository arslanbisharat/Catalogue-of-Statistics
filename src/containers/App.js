import { connect } from 'react-redux';
import { addHero, changeFilter } from '../actions/index';
import App from '../components/App';

const mapStateToProps = state => ({
  heroes: state.heroes,
  filter: state.filter,
  render: state.render,
});

const mapDispatchToProps = dispatch => ({
  addNewHero: hero => {
    dispatch(addHero(hero));
  },
  chngFilter: filter => {
    dispatch(changeFilter(filter));
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(App);
