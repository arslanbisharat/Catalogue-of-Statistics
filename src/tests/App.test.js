import React from 'react';
import {
  render, fireEvent, screen,
} from '@testing-library/react';
import '@testing-library/jest-dom';
import App from './App';
import initialState from './initialState';

beforeEach(() => {
  delete window.location;
  window.location = new URL('http://localhost/');
});

describe('Render the page correctly', () => {
  test('Render correctly', () => {
    const { asFragment } = render(<App />);
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('Open the application, click on the button e redirect to "HeroesList" page', () => {
  test('Render the application home page', () => {
    const { container } = render(<App />);
    expect(container.innerHTML).toContain('Welcome to the world of Comics');
  });

  test('Click on the button and render "HeroesList" page', () => {
    const { container } = render(<App />);
    fireEvent.click(screen.getByTestId('home-button'));
    expect(container.innerHTML).toContain('heroes-list');
  });

  test('The rendered page is the "Justice League of America" team page', () => {
    const { container } = render(<App />);
    fireEvent.click(screen.getByTestId('home-button'));
    expect(container.innerHTML).toContain('Justice League of America');
  });

  test('This is not the home page anymore', () => {
    const { container } = render(<App />);
    fireEvent.click(screen.getByTestId('home-button'));
    expect(container.innerHTML).not.toContain('Welcome to the world of Comics');
  });

  test('Test if the heroes data is being loaded - Superman', () => {
    const { container } = render(<App />);
    fireEvent.click(screen.getByTestId('home-button'));
    expect(container.innerHTML).toContain('<h2 class="hero-name">Superman</h2>');
  });

  test('Test if the heroes data is being loaded - Wonder Woman', () => {
    const { container } = render(<App />);
    fireEvent.click(screen.getByTestId('home-button'));
    expect(container.innerHTML).toContain('<h2 class="hero-name">Wonder Woman</h2>');
  });

  test('Test if the heroes data is being loaded - Batman', () => {
    const { container } = render(<App />);
    fireEvent.click(screen.getByTestId('home-button'));
    expect(container.innerHTML).toContain('<h2 class="hero-name">Batman</h2>');
  });
});

describe('Render the "HeroesList" page and click on the header`s buttons', () => {
  it('Renders the page and check the header', () => {
    const { container } = render(<App />);
    fireEvent.click(screen.getByTestId('home-button'));
    expect(container.innerHTML).toContain('Justice League of America');
  });

  it('Tests the "About" page', () => {
    const { container } = render(<App />);
    fireEvent.click(screen.getByTestId('home-button'));
    fireEvent.click(screen.getByText('About this project'));
    expect(container.innerHTML).toContain('<h1 class="about-title">About</h1>');
  });

  it('The About page does not have the home page title', () => {
    const { container } = render(<App />);
    fireEvent.click(screen.getByTestId('home-button'));
    fireEvent.click(screen.getByText('About this project'));
    expect(container.innerHTML).not.toContain('Welcome to the world of Comics');
  });

  it('The about page is not a team page', () => {
    const { container } = render(<App />);
    fireEvent.click(screen.getByTestId('home-button'));
    fireEvent.click(screen.getByText('About this project'));
    expect(container.innerHTML).not.toContain('Justice League of America');
  });

  it('Renders the "TeamsList" page and check the first title', () => {
    const { container } = render(<App />);
    fireEvent.click(screen.getByTestId('home-button'));
    fireEvent.click(screen.getByText('Click here to change filter'));
    expect(container.innerHTML).toContain('<h1>Filter subject by alignment:</h1>');
  });

  it('Renders the "TeamsList" page and check the second title', () => {
    const { container } = render(<App />);
    fireEvent.click(screen.getByTestId('home-button'));
    fireEvent.click(screen.getByText('Click here to change filter'));
    expect(container.innerHTML).toContain('<h1>Filter subject by filiation:</h1>');
  });

  it('Renders the "TeamsList" page and check the third title', () => {
    const { container } = render(<App />);
    fireEvent.click(screen.getByTestId('home-button'));
    fireEvent.click(screen.getByText('Click here to change filter'));
    expect(container.innerHTML).toContain('<h1>Choose a subject by name:</h1>');
  });

  it('Renders the "TeamsList" page and check for a heroe button', () => {
    const { container } = render(<App />);
    fireEvent.click(screen.getByTestId('home-button'));
    fireEvent.click(screen.getByText('Click here to change filter'));
    expect(container.innerHTML).toContain('Superman');
  });

  it('The TeamsList page is not the home page', () => {
    const { container } = render(<App />);
    fireEvent.click(screen.getByTestId('home-button'));
    fireEvent.click(screen.getByText('Click here to change filter'));
    expect(container.innerHTML).not.toContain('Welcome to the world of Comics');
  });
});

describe('Render the "TeamsList" page, click on a hero and then click on a team', () => {
  initialState.heroes.forEach(hero => {
    it(`Clicks on ${hero.name}'s button and render the "HeroesFile" page`, () => {
      const { container } = render(<App />);
      fireEvent.click(screen.getByTestId('home-button'));
      fireEvent.click(screen.getByText('Click here to change filter'));
      fireEvent.click(screen.getByText(hero.name));
      expect(container.innerHTML).not.toContain('<h1>Choose a subject by name:</h1>');
    });

    it('Check for the hero`s name', () => {
      const { container } = render(<App />);
      fireEvent.click(screen.getByTestId('home-button'));
      fireEvent.click(screen.getByText('Click here to change filter'));
      fireEvent.click(screen.getByText(hero.name));
      expect(container.innerHTML).toContain(hero.name);
    });

    it('Check for the hero`s identity', () => {
      const { container } = render(<App />);
      fireEvent.click(screen.getByTestId('home-button'));
      fireEvent.click(screen.getByText('Click here to change filter'));
      fireEvent.click(screen.getByText(hero.name));
      expect(container.innerHTML).toContain(hero.identity);
    });

    it('Check for the hero`s alignment', () => {
      const { container } = render(<App />);
      fireEvent.click(screen.getByTestId('home-button'));
      fireEvent.click(screen.getByText('Click here to change filter'));
      fireEvent.click(screen.getByText(hero.name));
      expect(container.innerHTML).toContain(hero.alignment);
    });

    it('Check for the hero`s filiation', () => {
      const { container } = render(<App />);
      fireEvent.click(screen.getByTestId('home-button'));
      fireEvent.click(screen.getByText('Click here to change filter'));
      fireEvent.click(screen.getByText(hero.name));
      expect(container.innerHTML).toContain(hero.filiation[0]);
    });
  });
});
