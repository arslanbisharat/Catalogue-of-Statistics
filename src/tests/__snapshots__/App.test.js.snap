// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render the page correctly Render correctly 1`] = `
<DocumentFragment>
  <div
    class="App"
    data-testid="App"
  >
    <div
      class="App"
    >
      <div
        id="init"
      >
        <h1
          class="app-title"
        >
          Welcome to the world of Comics
        </h1>
        <a
          href="/heroesList"
        >
          <button
            class="app-button"
            data-testid="home-button"
            type="submit"
          >
            Click Here
          </button>
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
